rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,418 @@
+
+### Challenge-Response Joining
+
+TODO: Consider alternative join method names?
+
+We believe a new join method, `challenge`, can meet our needs and provide
+significantly more flexibility than today's `token` join method. This works by -
+in a sense - inverting the token joining procedure: bots generate an ED25519
+keypair, and the public key is copied to the server. The public key can be
+copied out-of-band, or bots can provide their public key on first join using a
+one-time use shared secret, much like today's `token` method.
+
+Once the public key has been shared, bots may then join by requesting a
+challenge from the Teleport Auth service and complete it by signing it with
+their private key. If successful, the bot is issued a renewable identity just as
+`token`-joined bots are today, and the bot will actively renew this identity for
+as long as possible.
+
+If the identity renewal fails at any point, bots may attempt to reauthenticate,
+and the Auth service can use predefined per-bot rules to decide if this specific
+bot is allowed to rejoin, including a rejoin counter and expiration date. If a
+rejoin is rejected, the bot's identity does not necessarily remain invalid: if
+server-side rules are adjusted, for example by increasing the token's rejoin
+limit, it can then rejoin without any client-side reconfiguration.
+
+This has several important differences to existing join methods:
+
+- Onboarding secrets are optional, and the secret exchange process may be
+  skipped if the `ProvisionToken` is configured with a public key directly.
+  Otherwise, joining bots authenticate with an onboarding secret to
+  automatically share their public key with the server.
+
+- When joining or rejoining, Teleport issues a challenge that the client must
+  solve. This is similar to TPM joining today, but backed by a local keypair
+  rather than (necessarily) a hardware token.
+
+- When a bot's identity expires, assuming it has some rejoin allocations left,
+  it can simply repeat the joining process to receive a fresh renewable
+  certificate.
+
+- If a bot exhausts its rejoining limit, it will not be able to fetch new
+  certificates, similar to today's behavior. However, this bot can be restored
+  without needing to generate a new identity: an admin user can edit the backing
+  `ProvisionToken` to increment `spec.challenge.rejoining.total_rejoins`. The
+  failed `tbot` instance can then retry the joining process, and it will
+  succeed.
+
+It otherwise functions similarly to `token`-joined bots today. It proves its
+identity - either via an onboarding secret or public key - to receive a
+renewable identity and renews it as usual for as long as possible. The
+generation counter is still used to detect identity reuse. When the internal
+identity expires, the bot loses access to resources (until it reauthenticates).
+
+#### Token Resource Example
+
+`challenge`-type tokens differ from other types in that they are intended to
```

Comment:
We'll definitely want to discourage folks explicitly in documentation from setting a resource level expiry if it may break the rejoin mechanism.

Reply:
I could see it still being useful: if I want a bot that only lasts for e.g. 1 month, it would be nice if it could expire automatically, though we'd need some additional logic during regular cert renewals to ensure the backing token still exists.

Reply:
> though we'd need some additional logic during regular cert renewals to ensure the backing token still exists.

Yeah I think so - either way we just need to make sure the behaviour is fairly sane/well documented. I think worst case scenario would be someone deleting a known-bad token thinking it'll disconnect the bot - and it still has access.